File: Hw4/src/python/mcts.py
import numpy as np
from mcts_classes import POMDP, HistoryNode
import pudb

# time horizon
T = 0

# Observation probabilities (state x action x observation -> probability)
O = np.zeros((2, 3, 3))
# TODO: add observation probabilities

# Reward (state x action -> reward)
R = np.zeros((2, 3))
# TODO: add reward

# Transition probabilities (state x action x nextState -> probability)
P = np.zeros((2, 3, 2))
# TODO: add transition probabilities


terminalActions = np.array(())
pomdp = POMDP(T, O, R, P, terminalActions)
history = []

rootNode = HistoryNode(None, history,
                       pomdp.getBeliefFromHistory(history, np.array((.5, .5))),
                       pomdp)

for i in range(int(1e5)):
    # select
    selected, state, rewards, alreadyExpanded = rootNode.select(None, [])

    # expand
    if not alreadyExpanded:
        expanded, state, immediateReward = selected.expand(state)
        rewards.append(immediateReward)
    else:
        expanded = selected

    # simulate
    rewards.append(expanded.rollout(state))

    # update
    expanded.update(rewards, len(rewards)-1)

rootNode.getBestPath().printHistory()
pudb.set_trace()

File: Hw4/src/matlab_template/mcts.m
% Horizon
T= ...

% Observation probabilities (state x action x observation -> probability)
O = zeros(2,3,3);

% Reward (state x action -> reward)
R = zeros(2,3)

% Transition probabilities (state x action x nextState -> probability)
P = zeros(2,3,2); 

pomdp = POMDP(T, O, R, P, terminalActions);
terminalActions = [];
history = [];

rootNode = HistoryNode([], history,  pomdp.getBeliefFromHistory(history, [0.5;0.5]), pomdp);
for i=1:1e5
    % 1.select
    [selected, state, rewards, alreadyExpanded] = rootNode.select([],[]);
    
    % 2.expand
    if ~alreadyExpanded       
        [expanded, state, immediateReward] = selected.expand(state);
        rewards(end+1) = immediateReward;
    else
        expanded = selected;
    end
    
    % 3.simulate
    rewards(end+1) = expanded.rollout(state);
    
    % 4.update
    expanded.update(rewards, numel(rewards));
end
